tscache: enforce minimum retention window policy in intervalSkl

nvb opened the pull request:
This change adds a retention policy to `intervalSkl` that it should never evict
pages that contain intervals within a certain window of recency. This will allow
the structure to exhibit the `MinRetentionWindow` semantics that `tscache.Cache`
implementations require.

We do this by replacing the `earlier` and `later` `sklPages` with a linked list
of `sklPages` that can dynamically grow when necessary.

See https://github.com/cockroachdb/cockroach/pull/19508#issuecomment-339337357 for more discussion.

nvb committed:
tscache: enforce minimum retention window policy in intervalSkl

This change adds a retention policy to `intervalSkl` that it should never evict
pages that contain intervals within a certain window of recency. This will allow
the structure to exhibit the `MinRetentionWindow` semantics that `tscache.Cache`
implementations require.

We do this by replacing the `earlier` and `later` `sklPages` with a linked list
of `sklPages` that can dynamically grow when necessary.

See https://github.com/cockroachdb/cockroach/pull/19508#issuecomment-339337357
for more discussion.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19929)
<!-- Reviewable:end -->

nvb committed:
tscache: limit concurrency in TestIntervalSklConcurrency race tests

Halving the number of goroutines we use in the test speeds it up
by ~4x. This helps avoid test time outs.

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though @tschottdorf and @andy-kimball should take a look as well.

What default are we going to use for the skip-list page size? 

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/tscache/interval_skl.go, line 148 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19929#-KyWGE64Ba-j_GJrqkcI:-KyWGE65LDb3EKIPFQ2t:bhfp860) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed6d1e0414bb141752385de06370a9a5e0c9c01e/pkg/storage/tscache/interval_skl.go#L148)):*
> ```Go
> 	// policy. As such, the minimum retention policy takes priority over the
> 	// maximum size policy (which is why we call this a "soft" maximum).
> 	size uint32
> ```

Perhaps rename to `softMaxSize`.

---

*[pkg/storage/tscache/interval_skl.go, line 163 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19929#-KyWGTKJhexKHgsZFyyZ:-KyWGTKJhexKHgsZFyy_:b-91241) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed6d1e0414bb141752385de06370a9a5e0c9c01e/pkg/storage/tscache/interval_skl.go#L163)):*
> ```Go
> 	// In order to ensure that timestamps never decrease, intervalSkl maintains
> 	// a floor timestamp, which is the minimum timestamp that can be returned by
> 	// the lookup operations. When the earlier page is discarded, its current
> ```

`s/earlier/earliest/g`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19929#-:-KyWHDIROXSvy_9e93H-:b-3ec3xu)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM, but once we've tested and stabilized the new TS cache we should probably have a debate about whether this potentially-unbounded growth is even something we want. In the long run we want every component to be able to regulate its memory usage; I think it would be best for the ts cache to have a fixed size even if that may limit the lifetime of transactions (or if we do want the ts cache to grow dynamically we have to identify something else we can dynamically shrink that won't lead to cascading problems).

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19929#-:-KyWcPhnlmwcRulklOTS:brh1ejq)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> I think it would be best for the ts cache to have a fixed size even if that may limit the lifetime of transactions (or if we do want the ts cache to grow dynamically we have to identify something else we can dynamically shrink that won't lead to cascading problems).

We can dynamically adjust the size of the RocksDB cache. With the large blocks of allocation in the new tscache impl it seems feasible to adjust the size of the RocksDB cache up and down in response to the tscache size.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19929#-:-KyWel2Ash_ggrgLEQBq:bsopo5h)*
<!-- Sent from Reviewable.io -->

nvb committed:
tscache: maintain page size instead of total size in intervalSkl

nvanbenschoten commented:
> What default are we going to use for the skip-list page size?

I was planning on making each page 64MB, which is the current size for the entire cache (but which seems small).

> or if we do want the ts cache to grow dynamically we have to identify something else we can dynamically shrink that won't lead to cascading problems

I'm not sure exactly how this works in other production systems, but at some point, I've always pictured us creating a system-wide memory policy for how memory bounds of different components grow, shrink, or stay constant in relation to each other as load increases. For lack of a better analogy, think of how flexbox bounds total size while allowing individual elements to dynamically adjust in size.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/storage/tscache/interval_skl.go, line 148 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19929#-KyWGE64Ba-j_GJrqkcI:-KyWk-2usJSdPVdElDtP:b-lofyqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed6d1e0414bb141752385de06370a9a5e0c9c01e/pkg/storage/tscache/interval_skl.go#L148)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Perhaps rename to `softMaxSize`.
</blockquote></details>

I actually switched this to `pageSize` because it's easier to think about.

---

*[pkg/storage/tscache/interval_skl.go, line 163 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19929#-KyWGTKJhexKHgsZFyyZ:-KyWjzqZw_6ExiDTE--s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed6d1e0414bb141752385de06370a9a5e0c9c01e/pkg/storage/tscache/interval_skl.go#L163)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/earlier/earliest/g`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19929#-:-KyWiYAqOnGzFFzkJ5PO:bvcf23e)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@cardbot this is similar to what's discussed [https://github.com/cockroachdb/cockroach/issues/15849#issuecomment-309504798](here) in the context of large transactions. (Pointing that out mostly to link these two issues).

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 2 of 2 files at r3, 1 of 1 files at r4, 2 of 2 files at r5.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19929#-:-KyYYddzY5Sd6SDZCP4u:b-ulwkav)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Thanks for the reviews.

> I suspect that eventually some kind of "span coalescing" will be needed, in which long-running transactions lay claim to larger and larger spans which take less memory to represent (but which also can block other transactions that overlap with those spans).

Are there benefits to this kind of span coalescing strategy (and also the lock promotion policy) other than reducing a transaction's footprint over time as it performs more and more operations? I ask because I've traditionally thought of this kind of retention policy as being in place for long running transactions, not necessarily for large transactions. I guess those two will often go hand in hand. Regardless, I agree with @tschottdorf that this kind of approach has applications in multiple areas of our transaction system and foresee them being used in the future.
